rust/src/index/vector/ivf.rs

```diff
@@ -129,7 +129,29 @@ impl IVFIndex {
 
 impl std::fmt::Debug for IVFIndex {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "Ivf({}) -> {:?}", self.metric_type, self.sub_index)
+        let partitions = self
+            .ivf
+            .lengths
+            .iter()
+            .enumerate()
+            .map(|(i, &len)| {
+                format!(
+                    "[{}]: length={}, centroid={:?}",
+                    i,
+                    len,
+                    self.ivf.centroids.value(i)
+                )
+            })
+            .collect::<Vec<_>>()
+            .join("\n");
+        write!(
+            f,
+            "Ivf({}) -> PQ={:?}, num_partitions={}, partition_info=\n{}",
+            self.metric_type,
+            self.sub_index,
+            self.ivf.num_partitions(),
+            partitions
+        )
```

Comment:
It would be nice to have something more structured and parseable, although if I understand the challenge here it's that we have many different possible types of indices. One possibility is to at least have them be guaranteed to be JSON. I'll suggest a way to do that below.

The other issue with this is this Debug implementation is going to flow up to anything that contains it. So when you debug format a dataset, it's going to print this for every index it has, which might be kind of verbose. So we should probably make this a separate method.

I think the easiest thing to do would be add a new method on the [Index trait](https://github.com/lancedb/lance/blob/60d261580597d7c9e21616cc5b22aefb81a780e7/rust/src/index.rs#L43-L46):

```rust
pub(crate) trait Index {
    ...

    fn statistics(&self) -> String;
}
```

Either way, the implementation for IVF looks like:

```rust
#[derive(Serialize)]
pub struct IvfIndexStats {
    metric_type: String,
    num_partitions: usize,
    partitions: Vec<IvfIndexPartitionStats>,
}

#[derive(Serialize)]
IvfIndexPartitionStats {
   ...
}

impl Index for IvfIndex {
    ...

    fn statistics(&self) -> Result<String> {
        serde_json::to_string(IvfIndexStats {
            metrics_type: format!("{}", self.metric_type),
            num_partitions: self.num_partitions,
            ...
        })
    }
}
```

We may in the future change the API so that `statistics()` returns a typed object, but I think returning a JSON string is probably good enough for now. On the Python side, this can be parsed into dictionaries. It it's really in demand, we can create a class hierarchy in Python later.

Reply:
That makes a lot of sense. I've used your suggestion.